src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,651 @@
-                        prev_offset
-                    }
-                    None => MzOffset::from(0),
-                };
-                offset
-                    .checked_sub(prev_offset)
-                    .map(move |diff| (pid, ts, diff))
-            }
-            _ => panic!("invalid binding"),
-        })
+pub struct ReclockFollower<FromTime: Timestamp, IntoTime: Timestamp> {
+    inner: Rc<RefCell<ReclockFollowerInner<FromTime, IntoTime>>>,
 }
 
-/// Integrates the provided updates containing the change of offset in the diff field into updates
-/// containing insertions and retractions of full offset statements.
-fn integrate<'a>(
-    prev_state: &'a HashMap<PartitionId, MzOffset>,
-    updates: &'a [(PartitionId, MzOffset)],
-) -> impl Iterator<Item = ((PartitionId, MzOffset), Diff)> + 'a {
-    updates.into_iter().flat_map(|(pid, diff)| {
-        let (retraction, prev_offset) = match prev_state.get(pid).copied() {
-            Some(prev_offset) => (Some(((pid.clone(), prev_offset), -1)), prev_offset),
-            None => (None, MzOffset::from(0)),
-        };
-
-        let next_offset = prev_offset + *diff;
-        retraction
-            .into_iter()
-            .chain([((pid.clone(), next_offset), 1)])
-    })
-}
-
-struct ReclockFollowerInner {
-    /// A dTVC trace of the remap collection containing all consolidated updates at
-    /// `t` such that `since <= t < upper` indexed by partition and sorted by time.
-    remap_trace: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+struct ReclockFollowerInner<FromTime: Timestamp, IntoTime: Timestamp> {
+    /// A dTVC trace of the remap collection containing all updates at `t: since <= t < upper`.
+    // NOTE(petrosagg): Once we write this as a timely operator this should just be an arranged
+    // trace of the remap collection
+    remap_trace: Vec<(FromTime, IntoTime, i64)>,
+    /// The frontier after which this operator must take correct reclock decisions
+    as_of: Antichain<IntoTime>,
     /// Since frontier of the partial remap trace
-    since: Antichain<Timestamp>,
+    since: Antichain<IntoTime>,
     /// Upper frontier of the partial remap trace
-    upper: Antichain<Timestamp>,
-    /// The upper frontier in terms of `SourceTime`. Any attempt to reclock messages beyond this
-    /// frontier will lead to minting new bindings.
-    source_upper: OffsetAntichain,
+    upper: Antichain<IntoTime>,
+    /// The since frontier in terms of `FromTime`. Any attempt to reclock messages stricly less
+    /// than this frontier will result in an error.
+    source_since: Antichain<FromTime>,
+    /// The upper frontier in terms of `FromTime`. Any attempt to reclock messages beyond this
+    /// frontier will result in an error.
+    source_upper: MutableAntichain<FromTime>,
+    /// A flag indicating whether the remap trace accumulates to correct values through the as_of
+    /// frontier.
+    initialized: bool,
```

Comment:
do we strictly need this flag, or does it make some errors better?

Reply:
I not strictly needed. I put this flag because we need to do a single compact call the first time we initialize. See the logic in `push_trace_batch`. I could alternatively check in there if the trace is initialized before applying the batch and only perform the compaction if the batch that was just pushed took us from uninitialized to initialized. I'll do that

Reply:
nice!